rules-tests/DowngradePhp73/Rector/ConstFetch/DowngradePhp73JsonConstRector/Fixture/add_throw_on_json_decode_error.php.inc

```diff
@@ -0,0 +1,99 @@
+<?php
+
+namespace Rector\Tests\DowngradePhp73\Rector\ConstFetch\DowngradePhp73JsonConstRector\Fixture;
+
+// Class
+class DowngradePhp73JsonConstRectorFixture
+{
+    public function method_json_decode(string $json): string | bool
+    {
+        try {
+            $payload = json_decode($json, null, 512, JSON_THROW_ON_ERROR);
+            return $payload;
+        } catch (\Exception $e) {
+            return false;
+        }
+    }
+
+    public function method_json_decode_multi_flags(string $json): string | bool
+    {
+        try {
+            $payload = json_decode($json, null, 512, JSON_OBJECT_AS_ARRAY | JSON_THROW_ON_ERROR);
+            return $payload;
+        } catch (\Exception $e) {
+            return false;
+        }
+    }
+}
+
+// Function
+function func_json_decode(string $json): string | bool
+{
+    try {
+        $payload = json_decode($json, null, 512, JSON_THROW_ON_ERROR);
+        return $payload;
+    } catch (\Exception $e) {
+        return false;
+    }
+}
+
+// Standalone
+$decode = json_decode($json, null, 512, JSON_THROW_ON_ERROR);
+
+?>
+-----
+<?php
+
+namespace Rector\Tests\DowngradePhp73\Rector\ConstFetch\DowngradePhp73JsonConstRector\Fixture;
+
+// Class
+class DowngradePhp73JsonConstRectorFixture
+{
+    public function method_json_decode(string $json): string | bool
+    {
+        try {
+            $payload = json_decode($json, null, 512, 0);
+            if (json_last_error() !== JSON_ERROR_NONE) {
+                throw new \Exception(json_last_error_msg());
+            }
+            return $payload;
+        } catch (\Exception $e) {
+            return false;
+        }
+    }
+
+    public function method_json_decode_multi_flags(string $json): string | bool
+    {
+        try {
+            $payload = json_decode($json, null, 512, JSON_OBJECT_AS_ARRAY);
+            if (json_last_error() !== JSON_ERROR_NONE) {
+                throw new \Exception(json_last_error_msg());
+            }
+            return $payload;
+        } catch (\Exception $e) {
+            return false;
+        }
+    }
+}
+
+// Function
+function func_json_decode(string $json): string | bool
+{
+    try {
+        $payload = json_decode($json, null, 512, 0);
+        if (json_last_error() !== JSON_ERROR_NONE) {
+            throw new \Exception(json_last_error_msg());
+        }
```

Comment:
this is incorrect, as it previously on `\Exception` to `return false`, so must be skipped instead to avoid regression, so basically if inside try catch, it must not add the if, here the example:

https://3v4l.org/k3U5V

with new behaviour, it throw exception, which should not.

Reply:
@samsonasik @TomasVotruba Not sure to follow you here, without the `if` statement nor the `JSON_THROW_ON_ERROR`, `json_encode()` / `json_decode()` does not raise any error, so the `catch` is not executed, that's exactly why this PR was made, you're actually removing the precise purpose of it.

The problem resides on the type of the exception thrown to match the one in the catch, you should not remove the if entirely or you defeat the purpose of this PR and we are back to square one.

In PHP > 7.3 that does not throw (initial behavior of `DowngradePhp73JsonConstRector`) : 

https://3v4l.org/1Y2Bv#v7.2.34

```php
try {
    $data = "\xB1\x31"; // Invalid UTF-8 sequence (binary data)
    $json = json_encode($data, 0);
    echo $json;
} catch (JsonException $e) {
    echo 'Error occurred during JSON encoding: ' . $e->getMessage();
}
```

While this replicate the same behavior as with `JSON_THROW_ON_ERROR` (goal of my PR) : 

https://3v4l.org/lb86Z#v7.2.34

```php
try {
    $data = "\xB1\x31"; // Invalid UTF-8 sequence (binary data)
    $json = json_encode($data, 0);
    if (json_last_error() !== JSON_ERROR_NONE) {
        throw new \Exception(json_last_error_msg());
    }
    echo $json;
} catch (Exception $e) {
    echo 'Error occurred during JSON encoding: ' . $e->getMessage();
}
```

Also the exception `JsonException` does not exist in PHP < 7.3 so your example is not valid in this context, the catch exception if a `JsonException` should be replaced by a standard `Exception`.

I could have worked that out if you didn't act so fast, with your last merge this PR is basically useless.

Reply:
See https://3v4l.org/k3U5V vs https://3v4l.org/gJkgE

Better approach to use exception stmts instead probably the real solution, but the quick fix is not add if to avoid regression.

Reply:
sorry, I don't get it, it's like you didn't read what I just said.

The 2 examples you just provided are not valid PHP 7.2.x (since `JsonException` does not exist for that version)

Reply:
Exception can be used, but need to map for what defined inside catch, or it will fatal error that mean regression.

Reply:
The fatal error is linked to the fact that `JsonException` does not exist in PHP < 7.3, we HAVE to throw an `Exception` here,  the problem will be in the `catch` anyway if it is not downgraded properly, you are overlooking the real issue here that is actually not from my PR.

So for this PR to actually do what it's supposed to do (and replicate the right behavior of `JSON_THROW_ON_ERROR` in PHP version < 7.3) we just need to also convert the `JsonException` in the `catch` by an `Exception` and that's it.

Reply:
No, this is on php 8.x https://3v4l.org/gJkgE

Reply:
From what I understood, this replacement/rule is not supposed to be applied on PHP > 7.2 but maybe I'm missing some behaviors of Rector (only using it since a few days) ? Why would it be applied on versions where `JSON_THROW_ON_ERROR` exists ?

Reply:
as a side note, this can be rework with more carefull approach, with check if catch Exception or Throwable, or JsonException directly, then map the code below catch inside if, then remove the try catch itself.

See my comment above https://github.com/rectorphp/rector-downgrade-php/pull/133#issuecomment-1600736075

Reply:
No worries, improvement is welcome, just need more carefull approach to avoid regression ;)

I have on list of downgrade improvement, one of them is patch invalid the docblock during downgrade, see https://github.com/rectorphp/rector/issues/8004 but that's not our priority right now as docblock doesn't cause fatal error ;)

Reply:
> No worries, improvement is welcome, just need more carefull approach to avoid regression ;)

btw, how could I have caught locally the regression issue you got before pushing the PR ?
Since all was green both locally and eventually in the CI, I don't know where to look for to reproduce it ?

Reply:
For this rule itself, for example, we have the following code:

```php
try {
    $data = "\xB1\x31"; // Invalid UTF-8 sequence (binary data)
    $json = json_encode($data, JSON_THROW_ON_ERROR);
    echo $json;
} catch (JsonException $e) {
    echo 'something is wrong';
}
```

Currently, in php 7.3, it output `something is wrong`, see https://3v4l.org/GkRKc#v7.3.33

When downgraded to php 7.2, it should keep show `something is wrong`, see https://3v4l.org/XSJ4W#v7.2.34

This should limit per use-case to avoid general change,eg: Try, then catch JsonException, no finally, no anything else, if that works, we can improve to another improvement :)